### ISSUE TITLE
Ai

### ISSUE THREAD

**Issue by majdisaafi21-collab** [majdisaafi21-collab/Majdi-#1]
"""
Emotion Bridge AI - Audio Emotion Classification
Uses pre-trained Wav2Vec2 and Hugging Face transformers to classify emotions
"""

import os
import sys
import warnings
import numpy as np
import librosa
import torch
from transformers import (
    Wav2Vec2Processor,
    HubertForSequenceClassification,
    pipeline
)
from typing import Dict, Tuple, List
import json

warnings.filterwarnings('ignore')

class EmotionBridgeAI:
    """
    Emotion Bridge AI - Classifies emotions from audio files
    Supports: Happy, Sad, Angry, Neutral
    """
    
    def __init__(self, model_name: str = "facebook/hubert-large-ls960-ft-emotion"):
        """
        Initialize the Emotion Classification model
        
        Args:
            model_name: Pre-trained model from Hugging Face Hub
        """
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        print(f"🔧 Using device: {self.device}")
        
        try:
            print(f"📦 Loading model: {model_name}")
            self.processor = Wav2Vec2Processor.from_pretrained(model_name)
            self.model = HubertForSequenceClassification.from_pretrained(model_name)
            self.model.to(self.device)
            self.model.eval()
            
            # Emotion labels mapping
            self.emotion_labels = {
                0: "Angry",
                1: "Disgust",
                2: "Fear",
                3: "Happy",
                4: "Neutral",
                5: "Sad",
                6: "Surprise"
            }
            
            print("✅ Model loaded successfully!\n")
        except Exception as e:
            print(f"❌ Error loading model: {e}")
            sys.exit(1)
    
    def load_audio(self, file_path: str, sample_rate: int = 16000) -> Tuple[np.ndarray, int]:
        """
        Load audio file and resample to target sample rate
        
        Args:
            file_path: Path to audio file (WAV format)
            sample_rate: Target sample rate (default: 16000 Hz)
            
        Returns:
            Tuple of (audio_array, sample_rate)
        """
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"❌ Audio file not found: {file_path}")
        
        if not file_path.lower().endswith('.wav'):
            raise ValueError("❌ Only WAV format is supported")
        
        try:
            print(f"🎵 Loading audio file: {file_path}")
            audio, sr = librosa.load(file_path, sr=sample_rate)
            duration = len(audio) / sample_rate
            print(f"✅ Audio loaded successfully - Duration: {duration:.2f}s\n")
            return audio, sr
        except Exception as e:
            print(f"❌ Error loading audio: {e}")
            raise
    
    def classify_emotion(self, audio_path: str) -> Dict[str, any]:
        """
        Classify emotion from audio file
        
        Args:
            audio_path: Path to WAV audio file
            
        Returns:
            Dictionary with emotion predictions and confidence scores
        """
        # Load audio
        audio, sr = self.load_audio(audio_path)
        
        # Process audio for model
        print("🔄 Processing audio...")
        inputs = self.processor(
            audio,
            sampling_rate=sr,
            return_tensors="pt",
            padding=True
        )
        
        # Move to device
        input_values = inputs["input_values"].to(self.device)
        
        # Get predictions
        print("🧠 Analyzing emotions...\n")
        with torch.no_grad():
            logits = self.model(input_values).logits
        
        # Calculate probabilities
        probabilities = torch.softmax(logits, dim=-1).cpu().numpy()[0]
        
        # Create results dictionary
        results = {
            "file": os.path.basename(audio_path),
            "predictions": {}
        }
        
        # Map emotions with confidence scores
        for idx, prob in enumerate(probabilities):
            emotion = self.emotion_labels.get(idx, f"Unknown_{idx}")
            results["predictions"][emotion] = float(prob)
        
        # Sort by confidence score (descending)
        sorted_predictions = sorted(
            results["predictions"].items(),
            key=lambda x: x[1],
            reverse=True
        )
        
        results["predictions"] = dict(sorted_predictions)
        results["primary_emotion"] = sorted_predictions[0][0]
        results["confidence"] = float(sorted_predictions[0][1])
        
        return results
    
    def display_results(self, results: Dict[str, any]) -> None:
        """
        Display emotion classification results in a formatted way
        
        Args:
            results: Results dictionary from classify_emotion()
        """
        print("=" * 60)
        print("🎭 EMOTION BRIDGE AI - RESULTS")
        print("=" * 60)
        print(f"\n📄 File: {results['file']}")
        print(f"\n🎯 Primary Emotion: {results['primary_emotion']}")
        print(f"📊 Confidence Score: {results['confidence']:.2%}\n")
        
        print("📈 Emotion Breakdown:")
        print("-" * 60)
        for emotion, confidence in results['predictions'].items():
            bar_length = int(confidence * 40)
            bar = "█" * bar_length + "░" * (40 - bar_length)
            print(f"{emotion:12} | {bar} | {confidence:6.2%}")
        print("-" * 60)
    
    def save_results(self, results: Dict[str, any], output_path: str = "emotion_results.json") -> None:
        """
        Save results to JSON file
        
        Args:
            results: Results dictionary
            output_path: Path to save JSON results
        """
        with open(output_path, 'w') as f:
            json.dump(results, f, indent=2)
        print(f"\n💾 Results saved to: {output_path}")

def main():
    """Main execution function"""
    
    # Example usage
    if len(sys.argv) > 1:
        audio_file = sys.argv[1]
    else:
        # Default test file
        audio_file = "sample_audio.wav"
        print(f"ℹ️  Usage: python emotion_bridge_ai.py <audio_file.wav>")
        print(f"📝 Using default: {audio_file}\n")
    
    # Initialize Emotion Bridge AI
    emotion_ai = EmotionBridgeAI()
    
    # Classify emotion
    results = emotion_ai.classify_emotion(audio_file)
    
    # Display results
    emotion_ai.display_results(results)
    
    # Save results
    emotion_ai.save_results(results)

if __name__ == "__main__":
    main()
